### DISCUSSION TITLE
how to avoid ViewModelViewHost fallingl back to Locate View without constract

### DISCUSSION THREAD

**Discussion by newbienewbie** [reactiveui/ReactiveUI#3712] [General]
## Senario

I have multiple different views that share the same ViewModel.
```c#
class ViewA : UserControl, IViewFor<WaitingFunVM>
{
}

class ViewB: UserControl, IViewFor<WaitingFunVM>
{
}

Locator.CurrentMutable.Register( typeof(ViewA), typeof<IVewFor<WaitingFunVM>>,  ConstractA);
Locator.CurrentMutable.Register( typeof(ViewB), typeof<IVewFor<WaitingFunVM>>,  ConstractB);
```
I only defined the ViewModel (e.g. `WaitingFunVM`)  and let those views defined by other developers. The other developers provide user-defined views in plugins. And finally I locate the related views by a contract:
```
<rxui:ViewModelViewHost x:Name="myview"  ViewContract="ConstractB" />
```
Normally, it will render the `ViewB` without any problem.

However, in some case, the `ViewB` implementation is optional: Plugins author might not implement the ViewB at all.

What I want is that it should show nothing instead of rending `ViewA` if a specify  `ViewContract="ConstractB"`.

## The Current Implementation

The [current implementation](https://github.com/reactiveui/ReactiveUI/blob/9c36b0f0701ee7005556ccafaeb503a96ff6b75f/src/ReactiveUI.Wpf/Common/ViewModelViewHost.cs#L163C31-L163C31) of `ViewModelViewHost` resolves the View in the following way:
```c#
  private void ResolveViewForViewModel(object? viewModel, string? contract)
  {
        // ...
        var viewInstance = viewLocator.ResolveView(viewModel, contract) ?? viewLocator.ResolveView(viewModel);
```
It means the current implementation falls back to resolve view with no `contract` if nothing found.

If the `ResolveViewForViewModel()` must not be `protect virtual`,  is it possible add a property so that we can bypass the fallback to `viewLocator.ResolveView(viewModel);`?

e.g. :
```c#
var viewInstance = viewLocator.ResolveView(viewModel, contract) ;
if(viewInstance == null && !this.ContractFallbackByPass) 
{
    viewInstance = viewLocator.ResolveView(viewModel);
}
```

**Comment by glennawatson** [comment]
Adding a property is definitely a possibility as long as it defaults to the current fallback technique so not to break existing users that rely on it.
